File: backend/src/datafusion/schemas/abuse.py
"""Pydantic schemas for abuse system API endpoints."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from datafusion.models.abuse import ConsequenceSeverity, TargetType
from datafusion.models.consequence import TimeSkip
from datafusion.models.inference import ContentRating


class AbuseRoleRead(BaseModel):
    """Schema for reading abuse roles."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    role_key: str
    display_name: str
    description: str
    authorized_domains: str  # JSON string
    can_modify_data: bool
    created_at: datetime
    updated_at: datetime


class AbuseActionRead(BaseModel):
    """Schema for reading abuse actions."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    role_id: UUID
    action_key: str
    name: str
    description: str
    target_type: TargetType
    content_rating: ContentRating
    detection_chance: float = Field(ge=0.0, le=1.0)
    is_audit_logged: bool
    consequence_severity: ConsequenceSeverity
    created_at: datetime
    updated_at: datetime


class AbuseExecuteRequest(BaseModel):
    """Request to execute an abuse action."""

    role_key: str
    action_key: str
    target_npc_id: UUID


class AbuseExecuteResponse(BaseModel):
    """Response from executing an abuse action."""

    execution_id: UUID
    action_name: str
    target_name: str
    immediate_result: str = Field(description="What you see/learn immediately")
    data_revealed: dict | None = Field(default=None, description="The actual data you accessed")
    was_detected: bool
    detection_message: str | None = None
    warning: str | None = Field(default=None, description="Content warning if applicable")


class RealWorldParallel(BaseModel):
    """Real-world case parallel to illustrate consequences."""

    case: str
    summary: str
    source: str


class ConsequenceChain(BaseModel):
    """Chain of consequences over time for an abuse execution."""

    execution_id: UUID
    time_skips_available: list[TimeSkip] = Field(
        description="Time periods that have consequence data"
    )
    current_time_skip: TimeSkip
    events: list[str] = Field(description="Events that occurred at this time skip")
    victim_impact: str | None = Field(default=None, description="How the victim was affected")
    victim_statement: str | None = Field(
        default=None, description="Direct quote from victim (if available)"
    )
    your_status: str = Field(description="Player's employment/legal status")
    real_world_parallel: RealWorldParallel | None = None

File: frontend/src/types/abuse.ts
/**
 * Type definitions for abuse simulation system
 */

import { ContentRating } from './npc';

export enum TargetType {
  ANY_NPC = 'any_npc',
  SPECIFIC_NPC = 'specific_npc',
  SELF = 'self',
}

export enum ConsequenceSeverity {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  SEVERE = 'severe',
  EXTREME = 'extreme',
}

export enum TimeSkip {
  IMMEDIATE = 'immediate',
  ONE_WEEK = '1_week',
  ONE_MONTH = '1_month',
  SIX_MONTHS = '6_months',
  ONE_YEAR = '1_year',
}

export interface AbuseRole {
  id: string;
  role_key: string;
  display_name: string;
  description: string;
  authorized_domains: string; // JSON string
  can_modify_data: boolean;
  created_at: string;
  updated_at: string;
}

export interface AbuseAction {
  id: string;
  role_id: string;
  action_key: string;
  name: string;
  description: string;
  target_type: TargetType;
  content_rating: ContentRating;
  detection_chance: number;
  is_audit_logged: boolean;
  consequence_severity: ConsequenceSeverity;
  created_at: string;
  updated_at: string;
}

export interface AbuseExecuteRequest {
  role_key: string;
  action_key: string;
  target_npc_id: string;
}

export interface AbuseExecuteResponse {
  execution_id: string;
  action_name: string;
  target_name: string;
  immediate_result: string;
  data_revealed: Record<string, unknown> | null;
  was_detected: boolean;
  detection_message: string | null;
  warning: string | null;
}

export interface RealWorldParallel {
  case: string;
  summary: string;
  source: string;
}

export interface ConsequenceChain {
  execution_id: string;
  time_skips_available: TimeSkip[];
  current_time_skip: TimeSkip;
  events: string[];
  victim_impact: string | null;
  victim_statement: string | null;
  your_status: string;
  real_world_parallel: RealWorldParallel | null;
}

export interface ContentWarning {
  warning_type: string;
  content_rating: ContentRating;
  description: string;
  appears_in: string[];
}

export interface ScenarioWarnings {
  scenario_key: string;
  scenario_name: string;
  description: string;
  warnings: ContentWarning[];
  can_filter_dark_content: boolean;
  educational_purpose: string;
}
